lib/internal/http2/core.js

```diff
@@ -171,33 +176,40 @@ function onSessionHeaders(id, cat, flags, headers) {
 // is a necessarily synchronous operation. We need to know immediately if
 // there are trailing headers to send.
 function onSessionTrailers(id) {
+  debug('checking for trailers');
   const streams = this._owner._state.streams;
   const stream = streams.get(id);
   // It should not be possible for the stream not to exist at this point.
   // If it does not exist, there is something very very wrong.
   assert(stream !== undefined,
          'Internal HTTP/2 Failure. Stream does not exist. Please ' +
          'report this as a bug in Node.js');
-  const trailers = Object.create(null);
-  stream.emit('fetchTrailers', trailers);
-  return mapToHeaders(trailers);
+
+  try {
+    const trailers = Object.create(null);
+    stream.emit('fetchTrailers', trailers);
+    return mapToHeaders(trailers, assertValidPseudoHeaderTrailer);
+  } catch (err) {
+    process.nextTick(() => stream.emit('error', err));
+  }
```

Comment:
`mapToHeaders` will throw synchronously is the input is invalid, I don't want that to bubble up to the native layer so I'm catching it here and emitting. We may want to eventually refactor `mapToHeaders` so that it can emit the error itself but I'd rather do that change separately.